### PR TITLE
feat(zk_toolbox): Allow running docker images for provers

### PR DESCRIPTION
**Pull request by Artemka374**
## What ❔

Add `--docker` flag for `zk_inception prover run` which will allow prover components to run from docker images.

## Why ❔

To decrease setup time/improve UX

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: zk_toolbox/crates/zk_inception/src/commands/prover/args/run.rs [hunk 1]
```diff
@@ -90,10 +113,18 @@ impl ProverRunArgs {
             .witness_vector_generator_args
             .fill_values_with_prompt(component)?;
 
+        let docker = self.docker.unwrap_or_else(|| {
+            Prompt::new("Do you want to run Docker image for the component?")
+                .default("false")
```
[the remark this answers is not in the document]
Comment (Artemka374) [reply zk_toolbox/crates/zk_inception/src/commands/prover/args/run.rs:212]: it doesn't allow to put boolean there...

File: zk_toolbox/crates/zk_inception/src/commands/prover/args/run.rs [hunk 2]
```diff
@@ -76,8 +187,14 @@ impl WitnessVectorGeneratorArgs {
     }
 }
 
+#[derive(Debug, Clone, Parser, Default)]
+pub struct FriProverRunArgs {
+    #[clap(long)]
+    pub max_allocation: Option<usize>,
```
Comment (sanekmelnikov) [inline zk_toolbox/crates/zk_inception/src/commands/prover/args/run.rs:193]: /// Memory allocation limit in bytes (for prover component)
    #[clap(long)]
    pub max_allocation: Option<usize>,

File: zk_toolbox/crates/zk_inception/src/commands/prover/run.rs [hunk 3]
```diff
@@ -29,114 +30,114 @@ pub(crate) async fn run(args: ProverRunArgs, shell: &Shell) -> anyhow::Result<()
...
+    path_to_prover: &PathBuf,
 ) -> anyhow::Result<()> {
-    logger::info(MSG_RUNNING_WITNESS_GENERATOR);
-    let config_path = chain.path_to_general_config();
-    let secrets_path = chain.path_to_secrets_config();
-    let round = args.round.expect(MSG_WITNESS_GENERATOR_ROUND_ERR);
+    let spinner = Spinner::new(&format!("Pulling image {}...", image_name));
+    let pull_cmd = Cmd::new(cmd!(shell, "docker pull {image_name}"));
+    pull_cmd.run().context(error)?;
+    spinner.finish();
```
Comment (sanekmelnikov) [inline zk_toolbox/crates/zk_inception/src/commands/prover/run.rs:114]: Isn't docker run going to do that anyway?

File: zk_toolbox/crates/zk_inception/src/commands/prover/run.rs [hunk 4]
```diff
@@ -29,114 +30,114 @@ pub(crate) async fn run(args: ProverRunArgs, shell: &Shell) -> anyhow::Result<()
...
+    logger::info(message);
 
-    let mut cmd = Cmd::new(cmd!(shell, "cargo run --release --bin zksync_witness_generator -- {round_str} --config-path={config_path} --secrets-path={secrets_path}"));
-    cmd = cmd.with_force_run();
-    cmd.run().context(MSG_RUNNING_WITNESS_GENERATOR_ERR)
-}
+    let mut cmd = Cmd::new(cmd!(
+        shell,
+        "docker run --net=host -v {path_to_prover}/data/keys:/prover/data/keys -v {path_to_prover}/artifacts:/artifacts -v {path_to_configs}:/configs {application_args...} {image_name} {args...}"
+    ));
```
Comment (sanekmelnikov) [inline zk_toolbox/crates/zk_inception/src/commands/prover/run.rs:121]: minor: we can do it as a follow up later, but can we adapt our [docker::run](https://github.com/matter-labs/zksync-era/blob/main/zk_toolbox/crates/common/src/docker.rs#L20) for that? ignore, if that's too much work.

**Review by sanekmelnikov — COMMENTED** [review]
Thanks for working on that!

File: zk_toolbox/crates/zk_inception/src/commands/prover/run.rs (hunk 4 above)
Comment (Artemka374) [reply zk_toolbox/crates/zk_inception/src/commands/prover/run.rs:121]: btw it's pretty tricky. Initially I tried to put all the docker args in the `application_args` but it seems that xshell puts args with `:` in a double quotes, which breaks the command.

**Review by Deniallugo — APPROVED** [review]

**Review by Deniallugo — DISMISSED** [review]

**Review by Deniallugo — APPROVED** [review]

**Review by artmakh — APPROVED** [review]

**Pull request merged**